vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,683 @@
+		// Importer: ,
+		Schema: map[string]*schema.Schema{
+			"cluster": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "ID of the vSphere cluster the host will belong to.",
+			},
+			"hostname": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "FQDN or IP address of the host.",
+			},
+			"username": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Username of the administration account of the host.",
+			},
+			"password": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Password of the administration account of the host.",
+				Sensitive:   true,
+			},
+			"thumbprint": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Host's certificate SHA-1 thumbprint. If not set then the CA that signed the host's certificate must be trusted.",
+			},
+			"license": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "License key that will be applied to this host.",
+			},
+			"force": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Force add the host to vsphere, even if it's already managed by a different vSphere instance.",
+				Default:     false,
+			},
+			"connected": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Set the state of the host. If set to false then the host will be asked to disconnect.",
+				Default:     true,
+			},
+			"maintenance": {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Description: "Set the host's maintenance mode. Default is false",
+				Default:     false,
+			},
+			"lockdown": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Set the host's lockdown status. Default is disabled. Valid options are 'disabled', 'normal', 'strict'",
+				Default:     "disabled",
+			},
+		},
+	}
+}
+
+func resourceVsphereHostRead(d *schema.ResourceData, meta interface{}) error {
+
+	// NOTE: Destroying the host without telling vsphere about it will result in us not
+	// knowing that the host does not exist any more.
+
+	// Look for host
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+	if hostID == "" {
+		return nil
+	}
+
+	// Find host and get reference to it.
+	hs, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		if soap.IsSoapFault(err) {
+			sf := soap.ToSoapFault(err)
+			_, ok := sf.Detail.Fault.(types.ManagedObjectNotFound)
+			if !ok {
+				log.Printf("[DEBUG] Error while searching host %s. Error: %s", hostID, err)
+				return err
+			}
+			d.SetId("")
+			return nil
+		}
+		log.Printf("[DEBUG] non SOAP error while searching host %s. Error %#v", hostID, err)
+		return err
+	}
+
+	maintenanceState, err := hostsystem.HostInMaintenance(hs)
+	if err != nil {
+		return err
+	}
+	d.Set("maintenance", maintenanceState)
+
+	// Retrieve host's properties.
+	log.Printf("[DEBUG] Got host %#v", hs)
```

Comment:
This can be ok, but can also be overly verbose to have the entire object. Might be better to just use the ID or name.

Reply:
Good call. Probably a left over from my tests.